### PR TITLE
make "Invalid Unit" error message more friendly

### PR DESCRIPTION
**Pull request by donbright**
While i was using Slint Pad i noticed it said "Invalid Unit" when i typed in em for a unit. I wasnt sure what units were valid. 

This will make the Invalid Unit error message say something like 

```
error: Invalid unit 'em'. Valid units are: %,phx,px,cm,mm,in,pt,rem,s,ms,deg,grad,turn,rad
```
Thanks

### CODE REVIEW

File: internal/compiler/literals.rs [hunk 1]
```diff
@@ -130,7 +133,13 @@ pub fn parse_number_literal(s: String) -> Result<Expression, String> {
         end += 1;
     }
     let val = s[..end].parse().map_err(|_| "Cannot parse number literal".to_owned())?;
-    let unit = s[end..].parse().map_err(|_| "Invalid unit".to_owned())?;
+    let unit = s[end..].parse().map_err(|_| {
+        format!(
+            "Invalid unit '{}'. Valid units are: {}",
+            s.get(end..).unwrap_or(&s),
+            Unit::iter().filter(|x| x.to_string().len() > 0).join(",")
```
Comment (tronical) [inline internal/compiler/literals.rs:140]:
```suggestion
            Unit::iter().filter(|x| x.to_string().len() > 0).join(", ")
```

File: internal/compiler/literals.rs [hunk 2]
```diff
@@ -154,10 +163,14 @@ fn test_parse_number_literal() {
...
     assert_eq!(doit("10000001phx"), Ok((10000001., Unit::Phx)));
 
-    let wrong_unit = Err("Invalid unit".to_owned());
     let cannot_parse = Err("Cannot parse number literal".to_owned());
-    assert_eq!(doit("10000001 phx"), wrong_unit);
     assert_eq!(doit("12.10.12phx"), cannot_parse);
-    assert_eq!(doit("12.12oo"), wrong_unit);
-    assert_eq!(doit("12.12€"), wrong_unit);
+
+    let valid_units = Unit::iter().filter(|x| x.to_string().len() > 0).join(",");
```
Comment (tronical) [inline internal/compiler/literals.rs:169]:
```suggestion
    let valid_units = Unit::iter().filter(|x| x.to_string().len() > 0).join(", ");
```

**Review by tronical — COMMENTED** [review]
I think that's a nice idea! Olivier, what do you think?

**Review by ogoffart — APPROVED** [review]

**Comment by ogoffart** [comment]
This now includes another change.  Could you perhaps make another PR for the simple examples? And keep this PR for the Invalid Unit?

If you're not sure how to do it with git, i can also manually merge the invalid unit patch.

**Comment by tronical** [comment]
@donbright Please see Olivier's [comment](https://github.com/slint-ui/slint/pull/2701#issuecomment-1549164243) and remove the commits that are unrelated. Or if you're not sure, I can also do that for you in a separate branch and merge that.

**Review by tronical — CHANGES_REQUESTED** [review]
@donbright Please see Olivier's [comment](https://github.com/slint-ui/slint/pull/2701#issuecomment-1549164243) and remove the commits that are unrelated. Or if you're not sure, I can also do that for you in a separate branch and merge that.

**Comment by ogoffart** [comment]
I've manually merged the first commit in [a255aaf](https://github.com/slint-ui/slint/commit/a255aaf8d40ddbf2ca7d0f00c8d1e0df2ebab63a)

Thanks for your patch

**Pull request closed without merging**